mixer/pkg/protobuf/yaml/dynamic/auth.go

```diff
@@ -0,0 +1,294 @@
+	if c.prefix == "" {
+		return map[string]string{
+			c.header: string(c.token),
+		}, nil
+	}
+	return map[string]string{
+		c.header: c.prefix + " " + string(c.token),
+	}, nil
+}
+
+func (c *headerToken) RequireTransportSecurity() bool {
+	return true
+}
+
+func loadCerts(caFile string) (*x509.CertPool, error) {
+	// load ca cert.
+	caCertPool := x509.NewCertPool()
+	// load public root cert.
+	publicCerts, err := ioutil.ReadFile(publicCACertsFile)
+	if err != nil {
+		log.Warnf("cannot load public root cert: %v", err)
+	} else {
+		caCertPool.AppendCertsFromPEM(publicCerts)
+	}
+
+	// load ca provided by file, such as istio CA certs.
+	if caFile != "" {
+		caCerts, err := ioutil.ReadFile(caFile)
+		if err != nil {
+			return nil, err
+		}
+		caCertPool.AppendCertsFromPEM(caCerts)
+	}
+	return caCertPool, nil
+}
+
+func getWhitelistSAN(cert []byte) []*url.URL {
+	var whitelistSAN []*url.URL
+	c, err := x509.ParseCertificate(cert)
+	if err != nil {
+		return whitelistSAN
+	}
+	for _, uri := range c.URIs {
+		whitelistSAN = append(whitelistSAN, uri)
+	}
+	return whitelistSAN
+}
+
+func buildMTLSDialOption(mtlsCfg *policypb.Mutual) ([]grpc.DialOption, error) {
+	// load peer cert/key.
+	pk := mtlsCfg.GetPrivateKey()
+	cc := mtlsCfg.GetClientCertificate()
+	peerCert, err := tls.LoadX509KeyPair(cc, pk)
+	if err != nil {
+		return nil, errors.Errorf("load peer cert/key error: %v", err)
+	}
+
+	// for simplicity, when using mtls, restrict out of process adapter cert subject to be mixer's spiffe identity
+	// TODO(bianpengyuan) make this configurable if needed
+	var whitelistSAN []*url.URL
+	if peerCert.Certificate != nil && len(peerCert.Certificate) > 0 {
+		whitelistSAN = getWhitelistSAN(peerCert.Certificate[0])
+	}
+
+	// Load certificates, include public certs and certs from config (such as istio CA certs).
+	caCertPool, err := loadCerts(mtlsCfg.GetCaCertificates())
+	if err != nil {
+		return nil, fmt.Errorf("failed to build TLS dial option: CA certificate cannot be loaded: %v", err)
+	}
+
+	customVerify := func(rawCerts [][]byte, _ [][]*x509.Certificate) error {
+		if os.Getenv("BYPASS_OOP_MTLS_VERIFICATION") == "true" {
+			log.Infof("BYPASS_OOP_MTLS_VERIFICATION=true - bypassing mtls custom verification")
+			return nil
+		}
+		certs := make([]*x509.Certificate, len(rawCerts))
+		for i, asn1Data := range rawCerts {
+			cert, err := x509.ParseCertificate(asn1Data)
+			if err != nil {
+				return errors.New("failed to parse certificate from server: " + err.Error())
+			}
+			certs[i] = cert
+		}
+		opts := x509.VerifyOptions{
+			Roots:         caCertPool,
+			CurrentTime:   time.Now(),
+			Intermediates: x509.NewCertPool(),
+		}
+
+		for i, cert := range certs {
+			if i == 0 {
+				// do not add leaf cert to intermediate certs
+				continue
+			}
+			opts.Intermediates.AddCert(cert)
+		}
+		if _, err = certs[0].Verify(opts); err != nil {
+			return nil
```

Comment:
Shouldn't this `return err`? It seems that you're returning success on failure.

Reply:
Done. Thanks for catching this! I also moved the location of env variable SAN check bypass logic as @kyessenov suggested to make customVerify() always verify cert chain.

Reply:
Could you add a test for that case? At the very least, we need a test to verify that we don't accept untrusted certificates, don't we?